<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Trevor D. Ruiz</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<link href="./favicon-lwr.png" rel="icon" type="image/png">
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/academicons-1.9.2/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/academicons-1.9.2/size.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-H3YKZX4J88"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-H3YKZX4J88', { 'anonymize_ip': true});
</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Trevor D. Ruiz</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="mailto:truiz01@calpoly.edu"> 
<span class="menu-text"><i class="fa-solid fa-envelope" aria-label="envelope"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://calendly.com/tdruiz"> 
<span class="menu-text"><i class="fa-solid fa-calendar-days" aria-label="calendar-days"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://maps.app.goo.gl/hsJ5LDByoa3ZfHrj9"> 
<span class="menu-text"><i class="fa-solid fa-location-dot" aria-label="location-dot"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://scholar.google.com/citations?user=YhlR1RkAAAAJ"> 
<span class="menu-text"><i class="ai  ai-google-scholar"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://orcid.org/0009-0001-1452-2542"> 
<span class="menu-text"><i class="ai  ai-orcid"></i></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ruizt"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://tdruiz.com"> 
<span class="menu-text"><i class="ai  ai-cv"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">




<section id="section" class="level3 page-columns page-full">
<h3></h3>
<div class="quarto-about-broadside column-page content">
  <div class="about-main">
    <div class="about-entity" style="background-image: url('img/portrait2.jpg')"></div>
    <div class="about-contents"><header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Trevor D. Ruiz</h1>
</div>
<div class="quarto-title-meta">
  </div>
</header><div id="heading">
<p>I’m an Assistant Professor in the Statistics Department at Cal Poly SLO. I’m interested in applied statistics, data analysis, collaborative science, and teaching.</p>
<section id="appointments" class="level2">
<h2 data-anchor-id="appointments">Appointments</h2>
<p><strong>California Polytechnic State University, San Luis Obispo</strong></p>
<p>Assistant Professor | 2023-present</p>
<p><strong>University of California, Santa Barbara</strong></p>
<p>Visiting Assistant Professor | 2020-2023</p>
</section>
<section id="education" class="level2">
<h2 data-anchor-id="education">Education</h2>
<p><strong>Oregon State University</strong></p>
<p>Ph.D.&nbsp;&amp; M.S. Statistics | 2015-2020</p>
<p><strong>Reed College</strong></p>
<p>B.A. Philosophy | 2007-2011</p>
</section>
</div></div>
  </div>
</div>

</section>
<section id="research" class="level1">
<h1>Research</h1>
<p>I’m interested in statistical methods for correlated and high-dimensional data with applications in life sciences. I also enjoy collaborative projects involving applied statistical modeling.</p>
<p>I’m eager to work with student collaborators and most of my research agenda is designed with this in mind. Often, this entails that my research projects focus on specific applications or data analyses that would be accessible to undergraduate students but also exhibit some methodological complexity potentially suitable for graduate-level research. If you’re interested in working with me, our would like to chat about what that might involve, please <a href="mailto:truiz01@calpoly.edu">get in touch</a>.</p>
<section id="current-projects" class="level2">
<h2 data-anchor-id="current-projects">Current projects</h2>
<section id="using-environmental-dna-edna-metabarcoding-to-identify-microbial-correlates-of-marine-mammal-abundance" class="level4">
<h4 data-anchor-id="using-environmental-dna-edna-metabarcoding-to-identify-microbial-correlates-of-marine-mammal-abundance">Using environmental DNA (eDNA) metabarcoding to identify microbial correlates of marine mammal abundance</h4>
<p><em>Collaboration with Erin Satterthwaite and others (Scripps Institute of Oceanography, UCSD)</em></p>
<p><em>Student collaborators (alphabetical order): Katherine Chan (MS Statistics); Gabrielle Low (BS Statistics); Nick Patrick (BS Statistics); Samantha Ward (BS Statistics)</em></p>
<p>This project seeks to identify microbes predictive of the presence and abundance of large marine mammals in the California current. The project utilizes 16S and 18S r-RNA metabarcoding of eDNA obtained from ocean water samples collected at multiple depths along transects in the California current, together with density estimates computed from visual sightings of common whale species recorded along those same transects. We formulate a seasonally-adjusted log-contrast-type model relating density estimates to the relative abundances of amplicon sequence variants (ASVs) and implement a sparse estimation procedure to identify small subsets of ASVs predictive of whale density from 2014-2020.</p>
</section>
<section id="modeling-baseline-metabolic-needs-of-pregnant-rattlesnakes-from-high-resolution-time-series-of-body-temperatures" class="level4">
<h4 data-anchor-id="modeling-baseline-metabolic-needs-of-pregnant-rattlesnakes-from-high-resolution-time-series-of-body-temperatures">Modeling baseline metabolic needs of pregnant rattlesnakes from high resolution time series of body temperatures</h4>
<p><em>Collaboration with Haley Moniz and Emily Taylor (Biological Sciences, Cal Poly)</em></p>
<p><em>Student collaborators (alphabetical order): Emma Reardon (BS Environmental Management &amp; Protection); Nicole Yee (BS Statistics)</em></p>
<p>This project seeks to explore the effect of pregnancy on metabolism among rattlesnakes. Beaupre &amp; Duvall (1998) measured oxygen consumption in male, non-pregnant female, and (early) pregnant female rattlesnakes and developed general equations for calculating resting metabolic rate given a snake’s body temperature and weight. This project will leverage high-resolution time series of body temperatures collected from pregnant and non-pregnant female rattlesnakes over a period of several months to estimate metabolic rate fluctuations over time using the above-mentioned framework, and develop exploratory and statistical analyses aimed at (1) identifying metabolic differences between pregnant and non-pregnant rattlesnakes and (2) understanding temporal shifts in metabolism associated with seasonal and lifecycle factors.</p>
</section>
<section id="covariate-adjustment-in-sparse-partial-least-squares-models-for-high-dimensional-compositional-data" class="level4">
<h4 data-anchor-id="covariate-adjustment-in-sparse-partial-least-squares-models-for-high-dimensional-compositional-data">Covariate adjustment in sparse partial least squares models for high-dimensional compositional data</h4>
<p><em>Supported by the Research, Scholarly &amp; Creative Activities Program awarded by the Cal Poly Division of Research</em></p>
<p>Many ecological studies collect physical samples of water, soil, or other media in order to extract and sequence environmental DNA (eDNA), and often these samples are co-located space and time with additional survey measurements. Integrating such data provides a potential means of investigating relationships between eDNA composition and ecological and environmental features across space and time, and although there are a variety of statistical methods that capture relationships between compositional (<em>i.e.,</em> genomic) and non-compositional (<em>i.e.,</em> environmental/physical/biological) variables, few facilitate adjustment of inferred relationships for covariates or dependence among observations due to spatial and temporal sampling patterns, both of which are present among ecological data. This project aims to develop a partial least squares (PLS) framework that explicitly incorporates these adjustments.</p>
</section>
</section>
<section id="publications" class="level2">
<h2 data-anchor-id="publications">Publications</h2>
<p><em>Listed in reverse chronological order.</em></p>
<ol reversed="">
<li>
H. A. Moniz, J. H. Buck, H. L. Crowell, S. M. Goetz, <strong>T. D. Ruiz</strong>, S. M. Boback, E. N. Taylor (2024). High thermal quality rookeries facilitate high thermoregulatory accuracy in pregnant female rattlesnakes. <em>Journal of Thermal Biology.</em> [<a href="https://doi.org/10.1016/j.jtherbio.2024.103948">paper</a>]
</li>
<li>
<strong>T. D. Ruiz</strong>, S. Bhattacharyya, S. C. Emerson (2023). Sparse estimation of parameter support sets for generalized vector autoregressions by resampling and model aggregation. <em>Under review.</em> [<a href="https://doi.org/10.48550/arXiv.2307.09684">preprint</a>]
</li>
<li>
<strong>T. D. Ruiz</strong>, S. Bhattacharyya, S. C. Emerson (2022). A graphical sufficient condition for the stability of first-order log-linear Poisson generalized vector autoregressive processes. [<a href="https://dx.doi.org/10.2139/ssrn.4283477">preprint</a>]
</li>
<li>
A. M. E. Ojwang', <strong>T. D. Ruiz</strong>, S. Bhattacharyya, S. Chatterjee, P. S. Ojiambo, D. H. Gent (2021). A general framework for spatio-temporal modeling of epidemics with multiple epicenters: application to an aerially dispersed plant pathogen. <em>Frontiers in Applied Mathematics and Statistics.</em> [<a href="https://doi.org/10.3389/fams.2021.721352">paper</a>]
</li>
<li>
<strong>T. D. Ruiz</strong>, S. Bhattacharyya, M. Balasubramanian, K. Bouchard (2020). Sparse and low-bias estimation of high dimensional vector autoregressive models. <em>Proceedings of the 2nd Conference on Learning for Dynamics and Control</em> in <em>Proceedings of Machine Learning Research.</em> [<a href="https://proceedings.mlr.press/v120/ruiz20a.html">paper</a>]
</li>
<li>
M. Balasubramanian, <strong>T. D. Ruiz</strong>, B. Cook, M. Prabhat, S. Bhattacharyya, A. Shrivastava, K. E. Bouchard (2020). Scaling of union of intersections for inference of granger causal networks from observational data. <em>IEEE International Parallel and Distributed Processing Symposium.</em> [<a href="https://doi.org/10.1109/IPDPS47924.2020.00036">paper</a>]
</li>
<li>
<strong>T. D. Ruiz</strong> (2020). Estimation and sparse selection of conditional probability models for vector time series. <em>Oregon State University.</em> [<a href="https://ir.library.oregonstate.edu/concern/graduate_thesis_or_dissertations/k0698g13s">thesis</a>]
</li>
<li>
<strong>T. D. Ruiz</strong> &amp; C. L. Ellison (2020). Spatiotemporally coherent tensor decompositions for the analysis of trajectory data. <em>Engineer Research and Development Center</em> ERDC/GRL TN-20-3. [<a href="https://erdc-library.erdc.dren.mil/jspui/bitstream/11681/37355/1/ERDC-GRL%20TN-20-3.pdf">technical note</a>]
</li>
<li>
D. H. Gent, S. Bhattacharyya, <strong>T. D. Ruiz</strong> (2019). Prediction of spread and regional development of hop powdery mildew: a network analysis. <em>Phytopathology.</em> [<a href="https://doi.org/10.1094/PHYTO-12-18-0483-R">paper</a>]
</li>
</ol>
</section>
<section id="student-work" class="level2">
<h2 data-anchor-id="student-work">Student work</h2>
<ul>
<li><p>K. G. Chan (2024). Using plankton eDNA to estimate whale abundances off the California coast: data integration and statistical modeling. California Polytechnic State University. [<a href="https://digitalcommons.calpoly.edu/theses/2863">thesis</a>]</p></li>
<li><p>S. Rumsey, E. Ho, C. Zheng, N. Setiawan, J. Park, M. Elcheikhali, T. D. Ruiz, A. Kolbeinsson, E. J. Daza (2023). Identifying case onset points for early detection of influenza-like illness. <em>ACM SIGKDD’23 Southern California Data Science Day.</em> [<a href="https://drive.google.com/file/d/1mUdp1IfIl41d4oWxOL9ri-ZlG6rz1sf8/view?usp=share_link">poster abstract</a>]</p></li>
<li><p>L. Umsted, J. Liu, P. Trujillo, E. Burrell, L. B. Lancheros, T. D. Ruiz, E. Noi, E. Zhou, S. Dodge (2023). Understanding and modeling human mobility response to California wildfires. <em>Proceedings of 29th ACM SIGKDD Conference on Knowledge Discovery and Data Mining (SIGKDD’23)</em> [<a href="https://par.nsf.gov/servlets/purl/10501109">poster abstract</a>]</p></li>
<li><p>M. Gupta, A. Adams, E. Satterthwaite, T. D. Ruiz (2022). A scrollytelling primer on hypoxia: developing a data storytelling tool to communicate ocean observing data to California citizens. <em>CalCOFI Conference 2022: Innovative Techniques and Novel Applications of Time Series Data to Marine Resource Management.</em> Contributed presentation. [<a href="https://calcofi.io/hypoxia-story/">data storytelling</a>]</p></li>
</ul>
</section>
</section>
<section id="teaching" class="level1">
<h1>Teaching</h1>
<p>My teaching focuses on statistical methods, applied statistics, and data science. A full list of courses I’ve taught follows. For more recent courses, student-facing materials can be found on the course websites. I’m happy to share additional materials for instructional purposes on request.</p>
<section id="cal-poly" class="level2">
<h2 data-anchor-id="cal-poly">Cal Poly</h2>
<section id="stat218" class="level4">
<h4 data-anchor-id="stat218">[<a href="https://stat218.tdruiz.com">STAT218</a>] Applied statistics for life sciences</h4>
<p><em>Winter 2024, Spring 2024, Winter 2025, Spring 2025</em></p>
<p>Data collection and experimental design, descriptive statistics, confidence intervals, parametric and non parametric one and two-sample hypothesis tests, analysis of variance, correlation, simple linear regression, chi-square tests. Applications of statistics to the life sciences. Substantial use of statistical software.</p>
</section>
<section id="stat425" class="level4">
<h4 data-anchor-id="stat425">[<a href="https://stat425.tdruiz.com">STAT425</a>] Probability theory</h4>
<p><em>Fall 2023, Fall 2024</em></p>
<p>Basic probability theory, combinatorial methods, independence, conditional and marginal probability, probability models for random phenomena, random variables, probability distributions, distributions of functions of random variables, mathematical expectation, covariance and correlation, conditional expectation.</p>
</section>
</section>
<section id="uc-santa-barbara" class="level2">
<h2 data-anchor-id="uc-santa-barbara">UC Santa Barbara</h2>
<section id="pstat197" class="level4">
<h4 data-anchor-id="pstat197">[<a href="https://pstat197.github.io">PSTAT197</a>] Data science capstone</h4>
<p><em>Winter 2022, Spring 2022, Fall 2022, Winter 2023, Spring 2023</em></p>
<p>A preparatory course in data science topics and collaborative research skills (PSTAT197A/CMPSC190DD, offered fall term) and two experiential learning courses in which students work in teams on interdisciplinary projects sponsored by an external partner and advised by a technical specialist (PSTAT197B-C/CMPSC190DE-DF, offered winter and spring terms). Project teams establish a group contract outlining basic agreements pertaining to the collaboration and agree on outputs to produce for the sponsor by the end of the project – typically software, documentation, and a presentation. The course sequence culminates with a public showcase of student work.</p>
</section>
<section id="pstat100" class="level4">
<h4 data-anchor-id="pstat100">[<a href="https://pstat100.tdruiz.com">PSTAT100</a>] Data science concepts and analysis</h4>
<p><em>Spring 2021, Winter 2022, Spring 2022, Spring 2023</em></p>
<p>A hands-on introduction to data science intended for intermediate-level students from any discipline with some exposure to probability and basic computing skills, but few or no upper-division courses in statistics or computer science. The course introduces central concepts in statistics – such as sampling variation, uncertainty, and inference – in an applied setting together with techniques for data exploration and analysis. Applications emphasize end-to-end data analyses.</p>
</section>
<section id="pstat131" class="level4">
<h4 data-anchor-id="pstat131">[PSTAT131/231] Statistical machine learning</h4>
<p><em>Winter 2021, Spring 2021</em></p>
<p>Statistical Machine Learning is used to discover patterns and relationships in large data sets. Topics include: data exploration, classification and regression tress, random forests, clustering and association rules, building predictive models focusing on model selection, and model comparison and performance evaluation. Emphasis is on concepts, methods and data analysis, and students are expected to complete a significant class project, individual or team based, using real-world data.</p>
</section>
<section id="pstat126" class="level4">
<h4 data-anchor-id="pstat126">[PSTAT126] Regression analysis</h4>
<p><em>Summer 2021, Fall 2021</em></p>
<p>A first course in linear regression for statistics majors covering both theory and applications with a strong emphasis on statistical computing and visualization techniques in R using tidyverse programming. Tools for reproducible analysis are also introduced. The course covers the linear model in matrix form, parameter estimation and interpretation, residual diagnostics, statistical inference, case influence statistics and outlier detection, common data transformations, categorical variables and interaction terms, and model selection and prediction.</p>
</section>
<section id="pstat120b" class="level4">
<h4 data-anchor-id="pstat120b">[PSTAT120B] Probability and statistics</h4>
<p><em>Fall 2020, Winter 2021, Summer 2021, Fall 2021</em></p>
<p>Distribution of sample mean and sample variance; t, chi-squared and F distributions; summarizing data by statistics and graphs; estimation theory for single samples: sufficiency, efficiency, consistency, method of moments, maximum likelihood; hypothesis testing: likelihood ratio test; confidence intervals.</p>
</section>
</section>
<section id="oregon-state-university" class="level2">
<h2 data-anchor-id="oregon-state-university">Oregon State University</h2>
<section id="st351-352" class="level4">
<h4 data-anchor-id="st351-352">[ST351-352] Introduction to statistical methods</h4>
<p><em>Spring 2018, Summer 2018</em></p>
<p>Study designs, descriptive statistics, collecting and recording data, probability distributions, sampling distributions for means and proportions, hypothesis testing and confidence intervals for means and proportions in one- and two-sample inference, and chi-square tests.&nbsp;</p>



</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2024 Trevor Ruiz</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>